### PR TITLE
Send bank statement descriptors to payment intents

### LINKED ISSUE #4329: Invoice number on the bank statement
### Description
Hi folks, to be able to identify a bank transaction to a Woocommerce order, it should be possible to display an invoice number (and possibly also a customer number) on the customer's bank statement. 

 There is no other way to identify a transaction that the customer sees on his bank statement with an order from Woocommerce.

### Design
This options can be displayed as shortcodes (or variables etc.) under 
WooCommerce > Settings > Woocommerce Payments > "Transaction Preferences"

### Dev notes | Bank Statement
Since the characters are limited to 22, the variables/shortcodes should be very short or shouldn't even count

### PR DESCRIPTION
**Pull request by cesarcosta99**
Fixed #4329

#### Changes proposed in this Pull Request

The usage of the statement descriptor in transactions is being addressed in this PR. This is the final part of the #3625's work.

For some portion of the transactions, the regular statement descriptor is being added to the payment intents already. Here we're making sure the descriptor is passed to the payment intent and the shortened descriptor is being included to take place on card payments, if enabled.

It also includes a small fix for the Blocks checkout when using UPE, where the `wcpay_selected_upe_payment_type` was not being sent.

#### Testing instructions

Test the descriptors set are sent to the payment intent.

image

There's a vast amount of combinations for this test, please try to test as much as you can:
- UPE enabled and disabled.
- Classic checkout and Blocks checkout.
- Card, non-card payment method and express checkout.
- `Add customer order number to the bank statement` enabled and disabled.

Also consider the following:
1. When no `Full bank statement` and `Shortened customer bank statement` are provided, the payment intent should either contain no statement descriptor or have the `Full bank statement` descriptor set.
2. When `Full bank statement` is provided and `Add customer order number to the bank statement` is disabled, the payment intent should contain the `Full bank statement`'s value in the statement descriptor regardless the payment method chosen.
3. In card transactions (express checkout included), when `Add customer order number to the bank statement` is enabled and `Shortened customer bank statement` is provided, the payment intent should contain the `Shortened customer bank statement`'s value concatenated with the order number in the statement descriptor.

**The test:**
1. Set values for both `Full bank statement` and `Shortened customer bank statement`.
2. Go to the [shop page](http://localhost:8082/shop/) and add a product to the cart.
3. Go to checkout and complete the order using the scenarios described above.
4. Go to [Payments](https://dashboard.stripe.com/test/payments) in the Stripe dashboard, find the payment you've just made and click on it.
5. Check whether the statement descriptor value (under Payment details section) match the expected value according to the scenario being tested.

-------------------

- [ ] Added changelog entries to both `readme.txt` and `changelog.txt` (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

### CODE REVIEW

**Comment by cesarcosta99** [comment]
Waiting for #3625 getting merged before requesting review for this, since the work here depends and it's based on it.

**Comment by cesarcosta99** [comment]
Rebased `develop` to resolve conflicts.

File: includes/class-wc-payments-utils.php
Comment (ricardo) [inline includes/class-wc-payments-utils.php:650]: @cesarcosta99 I don't think we're officially applying `@since`/`@version` to the DocBlock in WCPay, besides leftovers from the Stripe extension (maybe?) 😅.

We even have a futuristic `@since 4.7.0` 🤖.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (ricardo) [inline includes/wc-payment-api/class-wc-payments-api-client.php:288]: Nit: The `(optional)` part makes me wonder if this is the only optional parameter, but it's not the case. I think we can dismiss the `(optional)` and change the type to `string|null`.

Also applies for `create_and_confirm_intention`.

**Review by ricardo — CHANGES_REQUESTED** [review]
I didn't finish testing, will do on Monday, but so far tests well.

Advancing code review. 👍

File: includes/class-wc-payment-gateway-wcpay.php
Comment (cesarcosta99) [reply includes/class-wc-payment-gateway-wcpay.php]: Thanks for suggesting that, it makes more sense. I've applied it in c342dcb. I only changed the return in the `elseif` to return `$full_statement_descriptor` as per our discussion below.

I went through all the testing with UPE enabled and seems to work fine, I'd appreciate if you can make another round of testing, specially with UPE, since we've changed the first condition (compared to the previous in the UPE gateway) to make the approach reusable.

File: includes/class-wc-payments-utils.php
Comment (cesarcosta99) [reply includes/class-wc-payments-utils.php]: 🤖 I've seen many references of `@since` across the project, but I thought it was also a good call to remove them because those methods are very specific, and are not widely used. Removed them in 6d3422e and 86b05a7.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (cesarcosta99) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: Applied in 106f141.

**Comment by cesarcosta99** [comment]
Rebased `develop` since #3625 got merged and there were conflicts.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (ricardo) [inline includes/class-wc-payment-gateway-wcpay.php:1114]: Now that you confirmed that the full statement descriptor comes directly from Stripe, perhaps we don't want to pass it to the payment intent, unless it needs to be changed.

The full statement descriptor should always be the same from the Stripe dashboard, correct?`https://dashboard.stripe.com/test/connect/accounts/{acct_id}/identity`

I suppose what differs from the Stripe plugin now is that we don't (or can't), set the full statement descriptor for the standard Stripe account directly, so that's why we allow a custom statement descriptor from the plugin settings (I didn't check this assumption).

Since we can set the express account statement descriptor in WCPay, we may not want to pass it along with the request, since it will always default to its full form anyway. We also may not want to do that to not risk providing an invalid statement descriptor somehow.

Now that I look at the Stripe docs more carefully, we may actually want to save the short statement descriptor in the account instead: https://stripe.com/docs/api/accounts/object#account_object-settings-card_payments-statement_descriptor_prefix

And then pass a [suffix](https://stripe.com/docs/api/payment_intents/object#payment_intent_object-statement_descriptor_suffix).

So in the end, this could be refactored like so:

```php
$statement_descriptor_suffix = $this->maybe_use_statement_descriptor_suffix( $order->get_payment_method(), $order );
```

```php
public function maybe_use_statement_descriptor_suffix( $payment_method, $order ) {
    $statement_descriptor_prefix           = $this->get_account_statement_descriptor_prefix();
    $is_short_statement_descriptor_enabled = 'yes' === $this->get_option( 'is_short_statement_descriptor_enabled' );

    if ( in_array( $payment_method, [ 'card', 'woocommerce_payments' ], true ) && $is_short_statement_descriptor_enabled && ! empty( $statement_descriptor_prefix ) ) {
        // TODO: Refactor this:
        return WC_Payments_Utils::get_dynamic_statement_descriptor( $order );
    }

    return null;
}
```

`get_account_statement_descriptor_prefix` would grab the [statement descriptor prefix](https://stripe.com/docs/api/accounts/object#account_object-settings-card_payments-statement_descriptor_prefix) from the account, similar to `get_account_statement_descriptor`, but probably without logging errors.

Then this function should only return the `* #123` part, which is the suffix.

And then we need to rename the additional parameter `statement_descriptor` to `statement_descriptor_suffix`.

Note: I didn't test the code above.

**Review by ricardo — CHANGES_REQUESTED** [review]
Now that I understand better how the full statement descriptor works, I believe we should make the short statement descriptor work the same way, and not try to save it under the plugin settings, but use the Stripe account for that, or at least explore this possibility, otherwise it can be hard to update this later and make it backwards compatible.

If we end up not having time to merge this with the statement descriptor prefix coming from the Stripe account, we could probably hide #3625 and merge this in the next WCPay version IMO (please check with others as well).

With the current implementation this tests well, despite one test detailed below. This is what I tested:

- **Short statement descriptor:**
  - non-UPE:
    - ✅ Blocks checkout.
    - ✅ Classic checkout.
    - ✅ Pay for order page.
  - UPE:
    - ✅ Blocks checkout.
    - ✅ Classic checkout.
    - ❌ Pay for order page.
  - ✅ Payment Request Button checkout.
  - ✅ Checkout with saved payment method.
- **Full statement descriptor:**
  - non-UPE:
    - ✅ Blocks checkout.
    - ✅ Classic checkout.
    - ✅ Pay for order page.
  - UPE:
    - ✅ Blocks checkout.
    - ✅ Classic checkout.
    - ✅ Pay for order page.
  - ✅ Payment Request Button checkout.
  - ✅ Checkout with saved payment method.

Reproduction steps for **❌ Pay for order page**:
- Make sure UPE is enabled and short statement descriptor is enabled.
- Go to **WooCommerce > Orders > Add order**.
- Add an item to the order, e.g Album.
- Click Create.
- Click "Customer payment page →".
- Checkout with a new credit card.
- Notice the full statement descriptor is used, instead of the short one.

File: includes/class-wc-payments-utils.php
Comment (cesarcosta99) [reply includes/class-wc-payments-utils.php]: Created #3982.

**Comment by cesarcosta99** [comment]
Just commenting to say I've got back to this and it's in progress.

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `3643` or branch name `update/send-statement-descriptors-on-pi` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=update/send-statement-descriptors-on-pi) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: e13cd2c361a090d11b2987029f8348522ec2924d
- Build time: 2023-07-25 21:00:03 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/class-wc-payment-gateway-wcpay.php
Comment (cesarcosta99) [reply includes/class-wc-payment-gateway-wcpay.php]: >perhaps we don't want to pass it to the payment intent, unless it needs to be changed.

Agreed. I tweaked the logic [here](https://github.com/Automattic/woocommerce-payments/pull/3643/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R2369-R2371) to only return it when payment method is not a card, which doesn't set a statement descriptor even if it's set on Stripe – that's why I also recommended testing a non-card payment method, such as SEPA.

>we may actually want to save the short statement descriptor in the account instead

I've tried that approach and it didn't play well with the regular and suffix statement descriptors, not sure why. I've tried to follow [Stripe docs](https://stripe.com/docs/account/statement-descriptors#dynamic) and sent different combinations as well as setting all of them together and the payment intent always ended up having the full statement descriptor set in it.

File: includes/wc-payment-api/models/class-wc-payments-api-intention.php
Comment (cesarcosta99) [reply includes/wc-payment-api/models/class-wc-payments-api-intention.php]: No, removed it in 81e755a.

**Review by cesarcosta99 — COMMENTED** [review]
@ricardo, sorry for making the PR more complex, but I've brought the work that I've done (and you have reviewed already) in #3625 into this PR. That had been reverted once this PR ended up not being merged. Since the two things are tied, I think it's better to have everything in a single PR rather than taking the risk of going over the same situation in case this is not merged soon enough.

>I believe we should make the short statement descriptor work the same way, and not try to save it under the plugin settings, but use the Stripe account for that, or at least explore this possibility, otherwise it can be hard to update this later and make it backwards compatible.

As I said [here](https://github.com/Automattic/woocommerce-payments/pull/3643#discussion_r1143822912), this didn't work well with the suffix approach. Another reason I came to not do this, is that even though we set it in the Stripe account, it's not visible in the account information in the Stripe Connect dashboard, neither viewing the Connect account as a regular account in the regular dashboard. IMO, that makes it hard to understand where this information comes from when working with it without any context.

Regarding the "Pay for order page" bug, I fixed it, however, I had to copy part of the logic to the JS once the payment processing takes place in the front-end in this case. Updating the payment intent (via PHP) after it's completed is not possible.

Can you please give it another round of testing?

**Comment by cesarcosta99** [comment]
Rebased `develop` to resolve conflicts.

@ricardo, I'm requesting another review from you since you've got context from previous reviews. Please consider [my last comment](https://github.com/Automattic/woocommerce-payments/pull/3643#pullrequestreview-1351182828) as well as my last replies from previous reviews. You might want to go through the testing instructions from #3625 too, since those changes got introduced in this PR.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (ricardo) [reply includes/class-wc-payment-gateway-wcpay.php]: @cesarcosta99 I did some testing and was able to set the short statement descriptor (aka `statement_descriptor_prefix`) in the Stripe account level and create a PaymentIntent using that.

1. Changed the server to also accept a "statement_descriptor_suffix" as an argument [here](https://github.com/Automattic/woocommerce-payments-server/blob/9c8fdef554a456ed25a8addbabe5f41cd10e77b5/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php#L151) and [here](https://github.com/Automattic/woocommerce-payments-server/blob/9c8fdef554a456ed25a8addbabe5f41cd10e77b5/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php#L234).
2. Manually updated my account using `curl` to have a card statement descriptor prefix:
```bash
curl https://api.stripe.com/v1/accounts/{WCPay_Dev_CONNECTED_ACCOUNT_ID} \
  -u "{WCPay_Dev_SECRET_KEY}:" \ # Remember the ":" at the end of the secret key
  -d "settings[card_payments][statement_descriptor_prefix]"="WOO.COM"
```
3. Add the `statement_descriptor_suffix` to the WCPay client intentions request to something like "ID: 1234". (Like what you did with `$request->set_statement_descriptor()`.

The only issue I encountered was that I needed to put a Latin character on the statement descriptor suffix, otherwise Stripe would throw an error:
> invalid_request_error - statement_descriptor_suffix
> The statement descriptor must contain at least one Latin character.

**Comment by ricardo** [comment]
@cesarcosta99 I didn't fully review the code yet, but considering [it's possible](https://github.com/Automattic/woocommerce-payments/pull/3643#discussion_r1275482569) to set the statement descriptor prefix at the account level, I guess that would be the best way to do it, given the full statement descriptor being already stored at the account level.

I think this would also simplify the implementation a little bit. We wouldn't need to store the statement descriptor prefix in the plugin settings nor handle the logic on whether to use the full statement or the short statement. We'd pass a `statement_descriptor_suffix` to the request in case the account has a `statement_descriptor_prefix` in cache and has a setting such as `is_statement_descriptor_suffix_enabled ` – I didn't test this with other payment methods, but Stripe might handle the logic of which statement descriptor to use depending on the payment method.

Of course, this would require a server change as well, where we enable saving the `statement_descriptor_prefix` ([ref](https://github.com/Automattic/woocommerce-payments-server/blob/9c8fdef554a456ed25a8addbabe5f41cd10e77b5/server/wp-content/rest-api-plugins/endpoints/wcpay/class-accounts-controller.php#L879)) and allow it to be retrieved by the client along with the existing `statement_descriptor` in the account cache.

> Another reason I came to not do this, is that even though we set it in the Stripe account, it's not visible in the account information in the Stripe Connect dashboard

I personally don't see this as a problem, since the statement descriptor prefix can be seen from the WCPay dashboard and account cache, but others might disagree.

Maybe we should ask the architects team on their thoughts about saving it under the plugin settings, or the Stripe account.

> invalid_request_error - statement_descriptor_suffix
The statement descriptor must contain at least one Latin character.

Regarding this error. It might be worth asking Stripe about what's the "ideal" suffix there, if not an order ID, and how to prefix the order ID with a Latin character so that this requirement can be fulfilled.

**Comment by ricardo** [comment]
Not sure if the input width used to be smaller at the time this PR was created, but maybe we should double-check this with design now, in order to avoid regression.

Also checking if the labels are accurate now.

| Before | After |
| -- | -- |
| Statement descriptor – Before | Statement descriptor – After |

**Comment by cesarcosta99** [comment]
Thanks for reviewing, @ricardo. I'm putting the PR on hold since I can't prioritize these changes at the moment. I'll address your questions and concerns once I get back to this PR.

**Review by ricardo — CHANGES_REQUESTED** [review]
This is just to prevent future Slack alerts.

**Comment by cesarcosta99** [comment]
Closing this pull request due to changes in Fractal priorities. More details [in the original issue](https://github.com/Automattic/woocommerce-payments/issues/4329#issuecomment-2128036233).

**Pull request closed without merging**
